File: backend/app/schemas/processed_dashboard.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

SalesBucket = Literal["pending", "possible", "strong", "dormant"]
FollowUpStatus = Literal["overdue", "due_today", "due_this_week", "no_followup", "future"]
AgingBucketKey = Literal["0_7_days", "8_14_days", "15_30_days", "31_60_days", "60_plus_days"]


class ProcessedDashboardQuoteRow(BaseModel):
    id: int
    quote_ref: str | None
    eworks_quote_id: int
    customer_name: str | None
    site_address: str | None
    quote_value: float | None
    processed_at: str | None
    days_since_processed: int
    days_in_current_bucket: int
    last_follow_up_at: str | None
    next_follow_up_at: str | None
    follow_up_status: FollowUpStatus
    sales_bucket: SalesBucket
    sales_note: str | None
    assigned_sales_name: str | None
    assigned_sales_email: str | None
    assigned_sales_user_id: str | None
    eworks_status: str | None
    eworks_status_name: str | None
    tags: list[str] = Field(default_factory=list)
    quote_detail_link: str


class ProcessedDashboardCategory(BaseModel):
    count: int
    value: float
    average_age_days: float
    overdue_followups: int
    quotes: list[ProcessedDashboardQuoteRow]


class ProcessedDashboardCategories(BaseModel):
    pending: ProcessedDashboardCategory
    possible: ProcessedDashboardCategory
    strong: ProcessedDashboardCategory
    dormant: ProcessedDashboardCategory


class ProcessedDashboardAgingBucket(BaseModel):
    count: int
    value: float


class ProcessedDashboardAging(BaseModel):
    days_0_7: ProcessedDashboardAgingBucket = Field(alias="0_7_days")
    days_8_14: ProcessedDashboardAgingBucket = Field(alias="8_14_days")
    days_15_30: ProcessedDashboardAgingBucket = Field(alias="15_30_days")
    days_31_60: ProcessedDashboardAgingBucket = Field(alias="31_60_days")
    days_60_plus: ProcessedDashboardAgingBucket = Field(alias="60_plus_days")

    model_config = {"populate_by_name": True}


class ProcessedDashboardFollowUpReminders(BaseModel):
    overdue: list[ProcessedDashboardQuoteRow]
    due_today: list[ProcessedDashboardQuoteRow]
    due_this_week: list[ProcessedDashboardQuoteRow]
    no_followup_set: list[ProcessedDashboardQuoteRow]


class SalespersonPerformanceRow(BaseModel):
    salesperson_name: str | None
    salesperson_email: str | None
    assigned_count: int
    pipeline_value: float
    strong_value: float
    accepted_count: int
    rejected_count: int
    conversion_rate: float
    overdue_followups: int
    average_days_to_close: float | None


class AcceptedRejectedTrendMonth(BaseModel):
    month: str
    accepted_count: int
    rejected_count: int
    accepted_value: float
    rejected_value: float


class MonthlyPipelineValueMonth(BaseModel):
    month: str
    new_processed_value: float
    active_pipeline_value: float
    strong_pipeline_value: float
    accepted_value: float
    rejected_value: float


class ProcessedDashboardTotals(BaseModel):
    processed_quotes: int
    pipeline_value: float
    strong_value: float
    dormant_quotes: int
    overdue_followups: int
    due_today_followups: int
    no_followup_set: int
    average_age_days: float
    conversion_rate: float
    accepted_count: int
    rejected_count: int
    accepted_value: float
    rejected_value: float


class ProcessedDashboardRead(BaseModel):
    totals: ProcessedDashboardTotals
    categories: ProcessedDashboardCategories
    aging: dict[str, ProcessedDashboardAgingBucket]
    follow_up_reminders: ProcessedDashboardFollowUpReminders
    salesperson_performance: list[SalespersonPerformanceRow]
    accepted_rejected_trend: list[AcceptedRejectedTrendMonth]
    monthly_pipeline_value: list[MonthlyPipelineValueMonth]


class SalesPipelinePatch(BaseModel):
    sales_bucket: SalesBucket | None = None
    sales_note: str | None = None
    assigned_sales_user_id: str | None = None
    assigned_sales_email: str | None = None
    assigned_sales_name: str | None = None
    last_follow_up_at: str | None = None
    next_follow_up_at: str | None = None


class SalesPipelineRead(BaseModel):
    id: str
    synced_quote_id: int | None
    eworks_quote_id: int
    quote_ref: str | None
    sales_bucket: SalesBucket
    sales_note: str | None
    assigned_sales_user_id: str | None
    assigned_sales_email: str | None
    assigned_sales_name: str | None
    processed_at: str | None
    last_follow_up_at: str | None
    next_follow_up_at: str | None
    bucket_changed_at: str | None
    accepted_at: str | None
    rejected_at: str | None
    closed_at: str | None

File: frontend/src/lib/processed-dashboard.ts
import { apiFetch } from "@/lib/api";

export type SalesBucket = "pending" | "possible" | "strong" | "dormant";
export type FollowUpStatus = "overdue" | "due_today" | "due_this_week" | "no_followup" | "future";

export type ProcessedDashboardQuote = {
  id: number;
  quote_ref: string | null;
  eworks_quote_id: number;
  customer_name: string | null;
  site_address: string | null;
  quote_value: number | null;
  processed_at: string | null;
  days_since_processed: number;
  days_in_current_bucket: number;
  last_follow_up_at: string | null;
  next_follow_up_at: string | null;
  follow_up_status: FollowUpStatus;
  sales_bucket: SalesBucket;
  sales_note: string | null;
  assigned_sales_name: string | null;
  assigned_sales_email: string | null;
  assigned_sales_user_id: string | null;
  eworks_status: string | null;
  eworks_status_name: string | null;
  tags: string[];
  quote_detail_link: string;
};

export type ProcessedDashboardCategory = {
  count: number;
  value: number;
  average_age_days: number;
  overdue_followups: number;
  quotes: ProcessedDashboardQuote[];
};

export type ProcessedDashboardTotals = {
  processed_quotes: number;
  pipeline_value: number;
  strong_value: number;
  dormant_quotes: number;
  overdue_followups: number;
  due_today_followups: number;
  no_followup_set: number;
  average_age_days: number;
  conversion_rate: number;
  accepted_count: number;
  rejected_count: number;
  accepted_value: number;
  rejected_value: number;
};

export type SalespersonPerformanceRow = {
  salesperson_name: string | null;
  salesperson_email: string | null;
  assigned_count: number;
  pipeline_value: number;
  strong_value: number;
  accepted_count: number;
  rejected_count: number;
  conversion_rate: number;
  overdue_followups: number;
  average_days_to_close: number | null;
};

export type ProcessedDashboard = {
  totals: ProcessedDashboardTotals;
  categories: Record<SalesBucket, ProcessedDashboardCategory>;
  aging: Record<string, { count: number; value: number }>;
  follow_up_reminders: {
    overdue: ProcessedDashboardQuote[];
    due_today: ProcessedDashboardQuote[];
    due_this_week: ProcessedDashboardQuote[];
    no_followup_set: ProcessedDashboardQuote[];
  };
  salesperson_performance: SalespersonPerformanceRow[];
  accepted_rejected_trend: Array<{
    month: string;
    accepted_count: number;
    rejected_count: number;
    accepted_value: number;
    rejected_value: number;
  }>;
  monthly_pipeline_value: Array<{
    month: string;
    new_processed_value: number;
    active_pipeline_value: number;
    strong_pipeline_value: number;
    accepted_value: number;
    rejected_value: number;
  }>;
};

export type SalesPipelinePatch = {
  sales_bucket?: SalesBucket;
  sales_note?: string | null;
  assigned_sales_user_id?: string | null;
  assigned_sales_email?: string | null;
  assigned_sales_name?: string | null;
  last_follow_up_at?: string | null;
  next_follow_up_at?: string | null;
};

export async function getProcessedDashboard(
  apiBase: "manager" | "admin",
  search?: string,
): Promise<ProcessedDashboard> {
  const params = new URLSearchParams();
  if (search?.trim()) params.set("search", search.trim());
  const qs = params.toString();
  const res = await apiFetch<ProcessedDashboard>(
    `/api/v1/${apiBase}/processed-dashboard${qs ? `?${qs}` : ""}`,
  );
  const data = res.data;
  if (!data?.totals || !data?.categories) {
    throw new Error("Invalid processed dashboard response");
  }
  return data;
}

export async function patchSalesPipeline(quoteId: number, patch: SalesPipelinePatch): Promise<void> {
  await apiFetch(`/api/v1/processed-quotes/${quoteId}/sales-pipeline`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });
}

export const BUCKET_LABELS: Record<SalesBucket, string> = {
  pending: "Pending",
  possible: "Possible",
  strong: "Strong",
  dormant: "Dormant",
};

export const FOLLOW_UP_LABELS: Record<FollowUpStatus, string> = {
  overdue: "Overdue",
  due_today: "Due today",
  due_this_week: "Due this week",
  no_followup: "No follow-up",
  future: "Scheduled",
};

export const AGING_LABELS: Record<string, string> = {
  "0_7_days": "0–7 days",
  "8_14_days": "8–14 days",
  "15_30_days": "15–30 days",
  "31_60_days": "31–60 days",
  "60_plus_days": "60+ days",
};
